rust/main/agents/relayer/src/msg/op_submitter.rs

```diff
@@ -342,86 +371,247 @@ async fn prepare_task(
+        {
+            batch
+        } else {
+            continue;
+        };
+
+        process_batch(
+            domain.clone(),
+            batch,
+            &mut prepare_queue,
+            &submit_queue,
+            &confirm_queue,
+            &metrics,
+        )
+        .await;
+    }
+}
+
+#[instrument(skip_all, fields(%domain))]
+#[allow(clippy::too_many_arguments)]
+async fn prepare_lander_task(
+    entrypoint: Arc<PayloadDispatcherEntrypoint>,
+    domain: HyperlaneDomain,
+    mut prepare_queue: OpQueue,
+    submit_queue: OpQueue,
+    confirm_queue: OpQueue,
+    max_batch_size: u32,
+    max_submit_queue_len: Option<u32>,
+    metrics: SerialSubmitterMetrics,
+    db: Arc<dyn HyperlaneDb>,
+) {
+    loop {
+        if apply_backpressure(&submit_queue, &max_submit_queue_len).await {
             continue;
         }
-        let mut task_prep_futures = vec![];
-        let op_refs = batch.iter_mut().map(|op| op.as_mut()).collect::<Vec<_>>();
-        for op in op_refs {
-            trace!(?op, "Preparing operation");
-            debug_assert_eq!(*op.destination_domain(), domain);
-            task_prep_futures.push(op.prepare());
+
+        let batch = if let Some(batch) = get_batch_or_wait(&mut prepare_queue, max_batch_size).await
+        {
+            batch
+        } else {
+            continue;
+        };
+
+        let batch_to_process = filter_ops_to_prepare(
+            entrypoint.clone(),
+            &prepare_queue,
+            &confirm_queue,
+            db.clone(),
+            batch,
+        )
+        .await;
+
+        process_batch(
+            domain.clone(),
+            batch_to_process,
+            &mut prepare_queue,
+            &submit_queue,
+            &confirm_queue,
+            &metrics,
+        )
+        .await;
+    }
+}
+
+/// This function checks the status of the payloads associated with the operations in the batch.
+/// If the payload is dropped or does not exist, the operation will go through prepare logic.
+/// If the payload is not dropped, the operation is pushed to the confirmation queue.
+/// If there are an error on retrieving payload ids or statuses for an operation, the operation
+/// will be returned to prepare queue.
+async fn filter_ops_to_prepare(
+    entrypoint: Arc<PayloadDispatcherEntrypoint>,
+    prepare_queue: &OpQueue,
+    confirm_queue: &OpQueue,
+    db: Arc<dyn HyperlaneDb>,
+    batch: Vec<QueueOperation>,
+) -> Vec<QueueOperation> {
+    let mut ops_to_prepare = vec![];
+    for op in batch.into_iter() {
+        let maybe_op_to_prepare = filter_single_op_to_prepare(
+            entrypoint.clone(),
+            prepare_queue,
+            confirm_queue,
+            db.clone(),
+            op,
+        )
+        .await;
+
+        if let Some(op_to_prepare) = maybe_op_to_prepare {
+            ops_to_prepare.push(op_to_prepare);
+        }
+    }
+    ops_to_prepare
+}
+
+async fn filter_single_op_to_prepare(
```

Comment:
this function has quite a lot of cases that we can simplify IMO. Since there are only two outcomes for an operation - 1) either submit to the confirm queue or, 2) keep preparing this operation - in this function I would only categorize what needs to be done by returning a bool, representing whether to confirm this operation. Then I would do the "sending to the confirm step / adding to the Vec of ops to prepare" in the`confirm_already_submitted_operations` function above. This function should probably be named `has_operation_been_confirmed` given that

Reply:
Done